Mathlib/Tactic/ComputeDegree.lean

```diff
@@ -0,0 +1,94 @@
+/-
+Copyright (c) 2023 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Mathlib.Data.Polynomial.Degree.Definitions
+
+/-!
+###  Simple lemmas about `natDegree` and `degree`
+
+The lemmas in this section deduce inequalities of the form `natDegree f ≤ d` and `degree f ≤ d`,
+using inequalities of the same shape.
+This allows a recursive application of the `compute_degree_le` tactic on a goal,
+and on all the resulting subgoals.
+-/
+
+open Polynomial
+
+namespace Mathlib.Tactic.ComputeDegree
+
+section mylemmas
+
+variable {R : Type _}
+
+section semiring
+variable [Semiring R]
+
+theorem add {a b : Nat} {f g : R[X]} (hf : natDegree f ≤ a) (hg : natDegree g ≤ b) :
+    natDegree (f + g) ≤ max a b :=
+(f.natDegree_add_le g).trans $ max_le_max ‹_› ‹_›
+
+theorem mul {a b : Nat} {f g : R[X]} (hf : natDegree f ≤ a) (hg : natDegree g ≤ b) :
+    natDegree (f * g) ≤ a + b :=
+natDegree_mul_le.trans $ add_le_add ‹_› ‹_›
+
+theorem pow {a : Nat} (b : Nat) {f : R[X]} (hf : natDegree f ≤ a) :
+    natDegree (f ^ b) ≤ b * a :=
+natDegree_pow_le.trans (Nat.mul_le_mul rfl.le ‹_›)
+
+theorem C_le (a : R) : natDegree (C a) ≤ 0 := (natDegree_C a).le
+theorem nat_cast_le (n : Nat) : natDegree (n : R[X]) ≤ 0 := (natDegree_nat_cast _).le
```

Comment:
BTW, do you handle `(2 : R[X])` (a.k.a. `OfNat.ofNat 2 : R[X]`) in your tactic?

Reply:
I think so.  It seems that special support in this case only touches `0, 1`:
```lean
example [Semiring R] : natDegree (0 : R[X]) ≤ 0 := by
  compute_degree_le

example [Semiring R] : degree (1 : R[X]) ≤ 0 := by
  compute_degree_le

example [Semiring R] : natDegree (2 : R[X]) ≤ 0 := by
  compute_degree_le

example [Semiring R] : degree (OfNat.ofNat 2 : R[X]) ≤ 0 := by
  compute_degree_le
```
all work with the current tactic.

However, your comment highlighted a bug:
```lean
example [Semiring R] : degree (OfNat.ofNat 0 : R[X]) ≤ 0 := by
  compute_degree_le

example [Semiring R] : degree (OfNat.ofNat 1 : R[X]) ≤ 0 := by
  compute_degree_le
/-
tactic 'apply' failed, failed to unify
  degree ↑?n ≤ 0
with
  degree (OfNat.ofNat 1) ≤ ?b
-/
```
both fail.  I'll investigate this in the other PR.

Reply:
These `OfNat` instances use `Zero` and `One`, not `Nat.cast`.

Reply:
I think that I fixed this in the main tactic by using `Expr.numeral?` and being more systematic.

I find the new coercion/numeral system somewhat confusing.  Also, I still think that the special support affects `0, 1` in this case, but from `2` onward it *appears* to be uniform.